content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md

```diff
@@ -0,0 +1,808 @@
+
+```c
+// Let's assume we have the following, rather contrived,
+// C++ program that wants to add 2 numbers, but wants to use
+// an `add` function from a relocatable object file supplied by
+// the user.
+//
+// Let's also assume that the add function in the user-supplied
+// relocatable object will reference a symbol named `MAGIC` in its
+// definition.
+
+const int MAGIC = 42;
+
+int main(int argc, char* argv[]) {
+    int a = 1;
+    int b = 2;
+    
+    // Read the path of the user supplied relocatable object.
+    string userSuppliedObjectPath = ...;
+    
+    // Initialize your JIT class that uses JIT linking under the hood.
+    JIT J;
+    
+    // Add the relocatable object to your JIT.
+    J.addObject(userSuppliedObjectPath);
+    
+    // Lookup the `add` function in the newly added JIT object.
+    // Once all symbolic references within the user supplied object
+    // are resolved, the content is fixed up and emitted to memory.
+    // And we can then get a pointer to the `add` function.
+    auto *add = (int(*)(int, int))J.lookup("add").getAddress();
+    
+    // At this point the symbolic reference to `MAGIC` in add's
+    // definition must have been resolved to the memory address
+    // of the constant `MAGIC` that we defined in this program.
+
+    // Run the add function found in the JIT module.
+    int result = add(a, b);
+}
+```
+That said, JIT linking by itself is not something that is very useful for an end user. JIT linking is an enabler
+for certain use-cases with pre-compiled languages (some use-cases exist for JIT-compiled languages as well[^2]).
+
+1. **JIT compilers** (think of something like the JIT compilation component of the [Java Hotspot VM](https://en.wikipedia.org/wiki/HotSpot_(virtual_machine)), but for a
+   statically compiled language)
+2. **Debugger expression evaluators** (such as the LLDB expression evaluator)
+3. **REPLs** (such as [Cling](https://root.cern/cling/) and the currently experimental[^3] [Clang-REPL](https://clang.llvm.org/docs/ClangRepl.html))
+4. **Standalone scripts** (such as the [Swift scripts](https://jblevins.org/log/swift), where the JIT linking is
+   used to add an [immediate](https://github.com/apple/swift/blob/main/lib/Immediate/Immediate.cpp#L228) mode to
+   the compiler, which runs your code in-place via a JIT, rather than compiling it)
+5. **Scriptable extensions** (think about running JIT’d code in the context of some existing app, allowing the app
+   to be extended by JIT’d code rather than precompiled plugins)
+
+While the above use cases may seem different, they are really the same — **JIT linking enables linking code into
+existing processes (that may or may not already contain state/context), in an ABI-compatible way.**
+
+### LLVM JITLink
+
+LLVM JITLink is a JIT linking implementation, in the form of a low-level library within the LLVM infrastructure. It provides primitives for:
```

Comment:
I see that it may not be inherently clear what I meant by low-level. What I really meant to say was that it is lower level than ORC, which is, sort of, the public API for JIT linking in LLVM. Perhaps I can be explicit here and say that instead, so that the meaning of "low-level" is clear.